include/exadg/operators/operator_base.cpp

```diff
@@ -2153,6 +2155,173 @@ OperatorBase<dim, Number, n_components>::evaluate_face_integrals() const
+      ++cell_batch_index)
+  {
+    unsigned int const n_filled_lanes =
+      matrix_free->n_active_entries_per_cell_batch(cell_batch_index);
+
+    for(unsigned int v = 0; v < n_filled_lanes; ++v)
+    {
+      auto const & cell = matrix_free->get_cell_iterator(cell_batch_index, v);
+
+      auto & local_dof_indices = all_dof_indices[cell_batch_index * vectorization_length + v];
+      if(is_mg)
+        cell->get_mg_dof_indices(local_dof_indices);
+      else
+        cell->get_dof_indices(local_dof_indices);
+
+      for(auto const & i : local_dof_indices)
+        weights[i] += 1.;
+    }
+  }
+  weights.compress(dealii::VectorOperation::add);
+
+  for(unsigned int i = 0; i < weights.size(); ++i)
+  {
+    if(weights.in_local_range(i))
+      weights[i] = 1. / std::sqrt(weights[i]);
+  }
+  weights.update_ghost_values();
+
+  dealii::SparseMatrixTools::restrict_to_full_matrices<SparseMatrix,
+                                                       dealii::DynamicSparsityPattern,
+                                                       Number>(tmp_matrix,
+                                                               dsp,
+                                                               all_dof_indices,
+                                                               overlapped_block_matrix);
+
+  // factorize and store cell matrices
+  matrices.resize(n_cells, CellMatrix(dofs_per_cell));
+  for(unsigned int cell_batch_index = 0; cell_batch_index < matrix_free->n_cell_batches();
+      ++cell_batch_index)
+  {
+    unsigned int const n_filled_lanes =
+      matrix_free->n_active_entries_per_cell_batch(cell_batch_index);
+
+    for(unsigned int v = 0; v < n_filled_lanes; ++v)
+    {
+      // get overlapped cell matrix
+      auto const & overlapped_cell_matrix =
+        overlapped_block_matrix[cell_batch_index * vectorization_length + v];
+
+      // store the cell matrix and renumber lexicographic if necessary
+      auto & lapack_matrix = matrices[cell_batch_index * vectorization_length + v];
+      if(is_dg)
+      {
+        lapack_matrix = overlapped_cell_matrix;
+      }
+      else
+      {
+        auto const & lex = matrix_free->get_shape_info().lexicographic_numbering;
+        for(unsigned int i = 0; i < dofs_per_cell; i++)
+          for(unsigned int j = 0; j < dofs_per_cell; j++)
+            lapack_matrix.set(i, j, overlapped_cell_matrix[lex[i]][lex[j]]);
+      }
+
+      // factorize the cell matrix
+      lapack_matrix.compute_lu_factorization();
+    }
+  }
+}
+
+template<int dim, typename Number, int n_components>
+void
+OperatorBase<dim, Number, n_components>::apply_inverse_as_matrices(VectorType &       dst,
+                                                                   VectorType const & src) const
+{
+  if(is_dg)
+    matrix_free->cell_loop(&This::cell_loop_apply_inverse_block_diagonal_matrix_based,
+                           this,
+                           dst,
+                           src);
```

Comment:
Is it correct that "Additive Schwarz"  always refers to single elements in the literature and not a batch of elements? In case of a batch of elements, additive Schwarz would be different from "block-Jacobi" also for DG. So I am not sure if we should implement it like this or actually assert that `is_dg==false` in case of additive Schwarz. @bergbauer How is it done for block-Jacobi, do we assert `is_dg==true` in that case? If so, it would probably make sense to do it similarly for additive Schwarz, i.e. only implement the continuous Galerkin case.

Reply:
Additive Schwarz can be element wise but can also batch several elements meaning it is a more general term. We could call the current class `ElementwiseAdditiveSchwarz` to make this more clear. I am looking into strategies for batching but this is not ready for master yet and will be an extension for the future.

>So I am not sure if we should implement it like this or actually assert that is_dg==false in case of additive Schwarz.

I am strongly against asserting additive Schwarz for DG. One main advantage of additive Schwarz is that you can use it as a Smoother on any multigrid level (CG or DG). With BlockJacobi as a smoother no c-transfer is currently possible. Sure, currently it is equivalent to Block Jacobi for DG, which is why I chose to use the optimized path here. We can also run though the general path for DG if you think this is clearer from a code design perspective.
Applying the factorization should be much more expensive than zeroing the vector and applying the weighting anyways.

Reply:
I think for a clearer design, the function `internal_compute_factorized_as_matrices()` should distinguish between DG and CG on the top level, just as the function `apply_inverse_as_matrices()`. It looks like weights are computed always, but are not used in the DG case, right? 

Whether we then remove one of the classes, merge them, rename `blockJacobi` or `AdditiveSchwarz`, would be a subsequent question in my opinion.

Reply:
I think we have two options here:

- we can currently distinguish between CG and DG on the top level, using the block-Jacobi path for DG and the general additive Schwarz path for CG because at the moment we only support cellwise additive Schwarz with weighting by contributions per DoF
- we run everything through the general additive Schwarz path for both CG and DG and for the `compute` and `apply` function

The current solution is a mix of the two options above because in the `compute` function the call `LAPACKMatrix::compute_lu_factorization` dominates the computational effort and it is only called once per solve. In the `apply` function, the dominant function call is `LAPACKMatrix::solve` which has lower complexity than the factorize call. Also `apply` is called in every iteration of a solver which makes optimization more important.

@peterrum Do you have an opinion on this topic as you thought about additive Schwarz extensively?

Reply:
ec537287bdb9a94abf48b3083a5443abca3ec87e uses the `BlockJacobi` path for DG